utils/run-clang-tidy.py

```diff
@@ -199,36 +477,110 @@ def apply_fixes(args, tmpdir):
     subprocess.call(invocation)
 
 
-def run_tidy(args, tmpdir, build_path, queue, lock, failed_files):
+# List of files with a non-zero return code.
+failed_files = []
+
+def run_tidy(tmpdir, build_path, quiet, queue, lock):
     """Takes filenames out of queue and runs clang-tidy on them."""
+    global failed_files
     while True:
-        name = queue.get()
-        invocation = get_tidy_invocation(name, args.clang_tidy_binary, args.checks,
-                                         tmpdir, build_path, args.header_filter,
-                                         args.extra_arg, args.extra_arg_before,
-                                         args.quiet, args.config)
-
-        proc = subprocess.Popen(
-            invocation, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-        output, err = proc.communicate()
-        if proc.returncode != 0:
-            failed_files.append(name)
+        name, command, dir = queue.get()
+        hash = None
+        hit = False
+        passed = True
+        output = ''
+        err = ''
+        tidy_error = False
+
+        invocation = tidy_common_args + get_tidy_file_args(name, tmpdir)
+
+        if cache:
+            try:
+                hash = pp_hash(name, command, dir)
+            except:
+                print('Preprocessing failed for {}, skipping cache lookup.'.format(name),
+                      file=sys.stderr)
+                hash = None
+            if hash:
+                with lock:
+                    try:
+                        hit, passed, output, err = cache.lookup(hash)
+                    except:
+                        print('Cache lookup error for {} with hash {}.'.format(name, hash),
+                              file=sys.stderr)
+                        traceback.print_exc()
+                        hit = False
+
+        if not hit:
+            proc = subprocess.Popen(
+                invocation, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            output_r, err_r = proc.communicate()
+            output = output_r.decode('utf-8')
+            err = err_r.decode('utf-8')
+            tidy_error = proc.returncode != 0
+            passed = len(output) == 0
+
         with lock:
-            sys.stdout.write(' '.join(invocation) +
-                             '\n' + output.decode('utf-8'))
-            if len(err) > 0:
-                sys.stdout.flush()
-                sys.stderr.write(err.decode('utf-8'))
+            if tidy_error:
+                failed_files.append(name)
+            sys.stdout.write('\n' + ' '.join(invocation) +
+                             '\n' + output)
+            sys.stdout.flush()
```

Comment:
It looks like the flushing does not work as expected here, some of the outputs are inter mixed, e.g. [here](https://github.com/tothxa/widelands/actions/runs/5051191348/jobs/9062765463#step:6:980). Just a notice, I do not think that this is a problem.
```
clang-tidy -p=/home/runner/work/widelands/widelands/build /home/runner/work/widelands/widelands/src/logic/game_settings.cc

clang-tidy -p=/home/runner/work/widelands/widelands/build /home/runner/work/widelands/widelands/src/logic/map_objects/map_object_type.cc
Cache hit as OK
Cache hit as OK
```

Reply:
I know. Commit [6ec20a4](https://github.com/widelands/widelands/pull/5922/commits/6ec20a4bd3eeebfa41551b2777e23a534e39e8ea) tried to address it, but didn't help.

Reply:
I've checked the outputs. Actually stdout doesn't really need the file name headers. Each reported error starts with the file name and line number, this is what `check_clang_tidy_results.py` uses. Although this means that errors in included headers are there too, but the clang-tidy command lines don't add much value, because they get lost in the "noise".

So I propose that only the stdout from the clang-tidy checks should be printed to the stdout of the script, and the filenames and the version and enabled checks info would go on stderr. The workflow would then don't even `tee` stdout, only redirect it to the log file. So the clang-tidy step would not show the found problems on the fly, which is the price of having a well organised displayed output.

Reply:
I changed it to print the invocation on stderr, so at least the cache reports will always come after it. Let's see how the output looks this way, whether we should redirect stderr to `tee` as well.

I can't decide about the initial version and checks report. I'd move them to stderr too, but then it looks somewhat strange that both stdout and stderr of `clang-tidy -version / -list-checks` goes to stderr. And checks were always part of the log, so I'm not sure about removing them for this reason either.